open-log/rfcs/0001-storage.md

```diff
@@ -0,0 +1,199 @@
+
+## Goals
+
+- Define a write API for appending entries to a key's log
+- Define a scan API for reading entries from a key's log
+
+## Non-Goals (left for future RFCs)
+
+- Active polling or push-based consumption
+- Retention policies
+- Checkpoints and cloning
+- Key-range scans
+
+## Design
+
+### Key Encoding
+
+SlateDB keys are a composite of the user key and a `u64` sequence number. A version prefix and record type discriminator provide forward compatibility.
+
+```
+Log Entry:
+  SlateDB Key:   | version (u8) | type (u8) | key (bytes) | sequence (u64) |
+  SlateDB Value: | record value (bytes) |
+```
+
+The initial version is `1`. The type discriminator `0x01` is reserved for log entries. Additional record types (e.g., metadata, indexes) may be introduced in future RFCs using different discriminators.
+
+This encoding preserves lexicographic key ordering, enabling key-range scans. Entries for the same key are ordered by sequence number.
+
+#### Variable-Length Key Considerations
+
+With variable-length keys, the boundary between key and sequence number is ambiguous during comparison. Two approaches to handle this:
+
+**Custom comparator** — SlateDB could support a custom comparator that treats the key and sequence number as separate components. This would enable correct ordering without encoding changes.
+
+**Scan filtering** — Without a custom comparator, scans may return entries from adjacent keys when key bytes happen to interleave with sequence number bytes. Results can be filtered by exact key match.
+
+In practice, users are likely to use fixed-length keys, which avoids this issue entirely.
+
+### Sequence Numbers
+
+Sequence numbers are assigned from a single counter that is maintained by the SlateDB writer and is incremented after every append. Each key's log is monotonically ordered by sequence number, but the sequence numbers are not contiguous—other keys' appends are interleaved in the global sequence.
+
+This approach simplifies ingestion by avoiding per-key sequence tracking. The trade-off is that sequence numbers do not reflect the count of entries within a key's log.
+
+If SlateDB supports multi-writer in the future, each writer would maintain its own sequence counter. This design assumes each key would still have a single writer—interleaving appends from multiple writers to the same key would break monotonic ordering within that key's log.
+
+### SST Representation
+
+OpenData-Log proposes two enhancements to SlateDB's SST structure to support efficient `scan` and `count` operations.
+
+#### Block Record Counts
+
+Each block entry in the SST index would include a cumulative record count:
+
+```
+Block Entry: | block_offset | cumulative_record_count | first_key |
```

Comment:
Is the cumulative record count the number of records in the log upto and including this block? Or the number of records cumulatively in the SST including this block? A short definition would be helpful.